### PR TITLE
Fixes Crystal 1.13 regression issue

### LINKED ISSUE #88: Specs are broken on Crystal master
Specs for habitat are failing on current Crystal master / nightly build. This was discovered in the ecosystem tests (https://github.com/crystal-lang/test-ecosystem/actions/runs/9306105007/job/25614556241).

The responsible code is this:

https://github.com/luckyframework/habitat/blob/5fd8d7b11bbfb82f28cddc817f010d76690e49f2/src/habitat.cr#L234-L238

It compares type membership based `id` which is based on the name representation instead of the type itself. Names can be ambiguous, i.e. there are multiple names for the same type. Here the problem is that the id `Nil` is different from `::Nil`.

The failure seems to be caused by a change in the compiler which fixes the id of global paths to include the leading `::` which had previsouly been missing (https://github.com/crystal-lang/crystal/pull/14490).

With `decl` as `foo : String?`, the expression `decl.type.types.map(&.id)` evaluates to `[String, Nil]` on Crystal 1.12.2. In Crystal master it's `[String, ::Nil]` (which is correct because `String?` is equivalent to `String | ::Nil`).

Simply changing the check to `includes?(::Nil.id)` seems logical, but that won't work because `::Nil` implicitly resolves to a `TypeNode`, not a `Path`. The path returned from `TypeNode#id` is always global, but it does not have a `::` prefix.

My initial understanding is that this code shouldn't have worked in the first place because it relied on the erroneous `id` generation that was fixed in the mentioned issue.

We have to analyze the compiler issue a bit more and it might be possible that another compiler change might fix this. So for now I'm just giving a heads up.
And maybe there's a chance to refactor this macro code so it won't break in the future if it was indeed only working because it relied on erroneous behaviour.

<details>
<summary>Spec output</summary>

```
    ........E.EFE..
   
   Failures:
   
     1) Habitat can check for missing settings
        Failure/Error: expect_raises(Habitat::MissingSettingError, %(settings.constant_setting = RandomClass)) do
   
          Expected Habitat::MissingSettingError with "settings.constant_setting = RandomClass", got #<Habitat::MissingSettingError: The 'this_can_be_nil' setting for FakeServer was nil, but the setting is required.
   
          Try this...
   
            FakeServer.configure do |settings|
              settings.this_can_be_nil = some_value
            end
          > with backtrace:
            # src/habitat.cr:12:3 in 'raise_if_missing_settings!'
            # spec/habitat_spec.cr:287:7 in '->'
            # /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/example.cr:50:13 in 'internal_run'
            # /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/example.cr:38:16 in 'run'
            # /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/context.cr:20:23 in 'internal_run'
            # /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/context.cr:344:14 in 'run'
            # /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/context.cr:20:23 in 'internal_run'
            # /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/context.cr:155:7 in 'run'
            # /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/dsl.cr:237:7 in 'execute_examples'
            # /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/dsl.cr:220:13 in '->'
            # /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/crystal/at_exit_handlers.cr:14:19 in 'run'
            # /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/crystal/main.cr:52:14 in 'exit'
            # /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/crystal/main.cr:47:5 in 'main'
            # /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/crystal/main.cr:129:3 in 'main'
            # /lib/x86_64-linux-gnu/libc.so.6 in '??'
            # /lib/x86_64-linux-gnu/libc.so.6 in '__libc_start_main'
            # /home/runner/.cache/crystal/crystal-run-spec.tmp in '_start'
            # ???
   
        # spec/habitat_spec.cr:286
   
     2) Habitat works with nilable types
   
          Nil assertion failed (NilAssertionError)
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/nil.cr:113:7 in 'not_nil!'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/nil.cr:109:3 in 'not_nil!'
            from spec/habitat_spec.cr:7:3 in 'this_can_be_nil'
            from spec/habitat_spec.cr:245:5 in '->'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/example.cr:50:13 in 'internal_run'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/example.cr:38:16 in 'run'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/context.cr:20:23 in 'internal_run'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/context.cr:344:14 in 'run'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/context.cr:20:23 in 'internal_run'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/context.cr:155:7 in 'run'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/dsl.cr:237:7 in 'execute_examples'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/dsl.cr:220:13 in '->'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/crystal/at_exit_handlers.cr:14:19 in 'run'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/crystal/main.cr:52:14 in 'exit'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/crystal/main.cr:47:5 in 'main'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/crystal/main.cr:129:3 in 'main'
            from /lib/x86_64-linux-gnu/libc.so.6 in '??'
            from /lib/x86_64-linux-gnu/libc.so.6 in '__libc_start_main'
            from /home/runner/.cache/crystal/crystal-run-spec.tmp in '_start'
            from ???
   
   
     3) Habitat can set and reset config using a block
   
          Nil assertion failed (NilAssertionError)
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/nil.cr:113:7 in 'not_nil!'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/nil.cr:109:3 in 'not_nil!'
            from spec/habitat_spec.cr:7:3 in 'this_can_be_nil'
            from spec/habitat_spec.cr:262:5 in '->'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/example.cr:50:13 in 'internal_run'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/example.cr:38:16 in 'run'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/context.cr:20:23 in 'internal_run'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/context.cr:344:14 in 'run'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/context.cr:20:23 in 'internal_run'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/context.cr:155:7 in 'run'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/dsl.cr:237:7 in 'execute_examples'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/dsl.cr:220:13 in '->'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/crystal/at_exit_handlers.cr:14:19 in 'run'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/crystal/main.cr:52:14 in 'exit'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/crystal/main.cr:47:5 in 'main'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/crystal/main.cr:129:3 in 'main'
            from /lib/x86_64-linux-gnu/libc.so.6 in '??'
            from /lib/x86_64-linux-gnu/libc.so.6 in '__libc_start_main'
            from /home/runner/.cache/crystal/crystal-run-spec.tmp in '_start'
            from ???
   
   
     4) Habitat can be converted to a Hash
   
          Nil assertion failed (NilAssertionError)
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/nil.cr:113:7 in 'not_nil!'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/nil.cr:109:3 in 'not_nil!'
            from spec/habitat_spec.cr:7:3 in 'this_can_be_nil'
            from spec/habitat_spec.cr:7:3 in 'to_h'
            from spec/habitat_spec.cr:308:5 in '->'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/example.cr:50:13 in 'internal_run'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/example.cr:38:16 in 'run'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/context.cr:20:23 in 'internal_run'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/context.cr:344:14 in 'run'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/context.cr:20:23 in 'internal_run'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/context.cr:155:7 in 'run'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/dsl.cr:237:7 in 'execute_examples'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/spec/dsl.cr:220:13 in '->'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/crystal/at_exit_handlers.cr:14:19 in 'run'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/crystal/main.cr:52:14 in 'exit'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/crystal/main.cr:47:5 in 'main'
            from /home/runner/work/_temp/crystal-nightly-nightly-undefined/share/crystal/src/crystal/main.cr:129:3 in 'main'
            from /lib/x86_64-linux-gnu/libc.so.6 in '??'
            from /lib/x86_64-linux-gnu/libc.so.6 in '__libc_start_main'
            from /home/runner/.cache/crystal/crystal-run-spec.tmp in '_start'
            from ???
   
   
   Finished in 103.82 milliseconds
   15 examples, 1 failures, 3 errors, 0 pending
   
   Failed examples:
   
   crystal spec spec/habitat_spec.cr:271 # Habitat can check for missing settings
   crystal spec spec/habitat_spec.cr:243 # Habitat works with nilable types
   crystal spec spec/habitat_spec.cr:259 # Habitat can set and reset config using a block
   crystal spec spec/habitat_spec.cr:296 # Habitat can be converted to a Hash
```
</details>

You can reproduce this with a [nightly build of Crystal](https://crystal-lang.org/install/nightlies/) (for example `docker run --rm -it -v $(pwd):/app -w /app crystallang/crystal:nightly-alpine crystal spec`).

### PR DESCRIPTION
**Pull request by jwoertink**
Fixes #88

See comments in #88 for a more in-depth context.

TL;DR We used to use `Nil.id` to check for nilable types, but the next Crystal release 1.13 will require `"::Nil".id`

/cc @straight-shoota

### CODE REVIEW

File: src/habitat.cr [hunk 1]
```diff
@@ -31,10 +31,17 @@ class Habitat
     # Habitat.raise_if_missing_settings!
     # ```
     def self.raise_if_missing_settings!
+      {% # https://github.com/crystal-lang/crystal/pull/14490
+
+ if compare_versions(Crystal::VERSION, "1.13.0-dev") >= 0
+   nil_type_id = "::Nil".id
+ else
+   nil_type_id = "Nil".id
+ end %}
```
Comment (jwoertink) [inline src/habitat.cr]: The formatting is weird.. This is from `crystal tool format`

File: src/habitat.cr [hunk 2]
```diff
@@ -226,12 +233,19 @@ class Habitat
         {% end %}
       {% end %}
 
+      {% # https://github.com/crystal-lang/crystal/pull/14490
+ if compare_versions(Crystal::VERSION, "1.13.0-dev") >= 0
```
Comment (jwoertink) [inline src/habitat.cr]: I need the `-dev` here otherwise the current nightly wouldn't get picked up. I'm hoping the final release is still `>=` the `-dev`.

File: src/habitat.cr [hunk 3]
```diff
@@ -31,10 +31,17 @@ class Habitat
     # Habitat.raise_if_missing_settings!
     # ```
     def self.raise_if_missing_settings!
+      {% # https://github.com/crystal-lang/crystal/pull/14490
+
+ if compare_versions(Crystal::VERSION, "1.13.0-dev") >= 0
+   nil_type_id = "::Nil".id
```
Comment (matthewmcgarvey) [inline src/habitat.cr]: I'm wary of this change because https://github.com/luckyframework/habitat/issues/88 seems like it's saying this might not work or, at least, work in every scenario.

I think I found a way to update this to work for current and future versions.

Like he said with path having the leading "::", if you call [`Path#names`](https://crystal-lang.org/api/1.12.2/Crystal/Macros/Path.html#names%3AArrayLiteral%28MacroId%29-instance-method) it should separate based on those and the leading "::" is dropped. `Namespaced::Class` becomes `[Namespaced, Class]` and `::Nil` is just `[Nil]`. To use that in this scenario, the condition becomes

```crystal
!setting[:decl].type.types.any?(&.names.includes?(Nil.id)))
```

Never used `names` before, but it seems like it works well and we don't need any crystal version conditionals.

Comment (jwoertink) [reply src/habitat.cr]: Oh, nice find! I don't think I've seen that either, but definitely worth a shot since this felt pretty hacky as is.

**Comment by jwoertink** [comment]
@matthewmcgarvey good catch! That is a much nicer fix, and seems to work :raised_hands:
